mantis/simulation/routers/lautaro.py

```diff
@@ -0,0 +1,213 @@
+        return self.U[0]
+
+class Estado:
+    max_depth : int
+    depth : int
+    dist : list[list[tuple[int, float]]]
+    dlock : list[th.Lock]
+    u_end : int
+    edges : list[Edge]
+    revision : bool
+    Nopts : int
+    j : int
+    def __init__(self):
+        self.dist = None
+        self.max_depth = None
+        self.depth = None
+        self.u_end = None
+        self.edges = None
+        self.revision = None
+        self.dlock = None
+        self.Nopts = None
+        self.j = 0
+    
+def Rango(e0, e1, estado):
+        edges = estado.edges
+        j = estado.j
+        #print(e0, e1, j)
+
+        dist = estado.dist
+        for ei in range(e0,e1):
+            e = edges[ei]
+            for u in e.U:
+                if dist[j][u][1] == 0: continue
+                v = e.GetOther(u)
+                if estado.revision:
+                    ee = copy.deepcopy(e)
+                    vv = u
+                    for jj in range(j,0,-1):
+                        pad = dist[jj][vv][0]
+                        vv = edges[pad].GetOther(vv)
+                        if pad == ei: 
+                            ee.DoChange(vv,dist[jj-1][vv][1])
+                else : ee = e
+                Xv = ee.GetAmount(u, dist[j][u][1])
+                estado.dlock[v].acquire()
+                if dist[j+1][v][1] < Xv:
+                    dist[j+1][v] = (ei, Xv )
+                estado.dlock[v].release()
+
+# Bellman Ford based solution
+
+def route(
+    input: Input,
+    all_data: AllData,
+    _ctx: Ctx = Ctx(),
+    max_depth:int = 5,
+    splits:int = 1000,
+    revision = True,
+    Nproces = None,
+):
+    edges : list[Edge] = []
+    all_tokens = all_data.all_tokens
+    tokensIds = {x:i for i,x in enumerate(all_tokens)}
+    
+    if Nproces == None: Nproces = os.cpu_count()
+
+    if isinstance(max_depth,int): max_depth = [max_depth]
+    if isinstance(splits,int): splits = [splits]
+
+    for x in all_data.asset_transfers:
+        edges.append(Edge(x, tokensIds))
+    for x in all_data.asset_pairs_xyk:
+        edges.append(Edge(x, tokensIds))
+    
+    n = len(all_tokens)
+    
+    deltas : list[float] = [0]*len(edges)
+    lambdas: list[float] = [0]*len(edges) 
+    paths : list[list[int]] = []
+    outcomes : list[float] = [0]
+    totSplits = sum(splits)
+
+    u_init = tokensIds[input.in_token_id]
+    u_end = tokensIds[input.out_token_id]
+
+    estado = Estado()
+    estado.u_end = u_end
+    estado.edges = edges
+    estado.revision = revision
+    estado.dlock = [th.Lock() for i in range(n)]
+    #print(estado.dlock)
+
+    e0 = [i*len(edges)//Nproces for i in range(Nproces)]
+    e1 = [(i+1)*len(edges)//Nproces for i in range(Nproces)]
+    e1[-1] = len(edges)
+    estado.dlock = [th.Lock() for i in range(n)]
+    
+    n0 = [i*n//Nproces for i in range(Nproces)]
+    n1 = [(i+1)*n//Nproces for i in range(Nproces)]
+    n1[-1] = n
+
+
+    for max_depth_i, splits_i in zip(max_depth, splits):
```

Comment:
May be I miss something. Split is not used recursively? That would mean that path of transfer first and split swap will not work?